priv/posts/security/20241212210148_epmd-public-exposure.md

```diff
@@ -0,0 +1,126 @@
+{
+  "title": "Exposed EPMD: A Hidden Security Risk for RabbitMQ and the BEAM Ecosystem",
+  "authors": ["The Security Working Group"],
+  "slug": "epmd-public-exposure",
+  "category": "security",
+  "tags": ["security", "epmd", "cybersecurity", "rabbitmq"],
+  "datetime": "2024-12-12T21:01:48.619713Z"
+}
+---
+EPMD, essential for Erlang and RabbitMQ clustering, is often exposed online—posing hidden security risks and requiring quick mitigation steps.
+---
+
+The Erlang Port Mapper Daemon (EPMD) is a built-in component that helps Erlang-based applications (including RabbitMQ) discover each other’s distribution ports for clustering. Although EPMD itself isn’t directly dangerous, its exposure on the public internet often signals that Erlang Distribution ports are also exposed. This creates a serious security risk: if attackers find these distribution ports, they can potentially join your cluster, run arbitrary code, and compromise your systems. Recent scans have revealed over 85,000 instances of publicly accessible EPMD, with roughly half associated with RabbitMQ servers.
+
+If left unsecured, exposed Erlang Distribution ports let attackers gain a foothold in your system. Fortunately, mitigation steps are straightforward: disable EPMD if you’re not clustering, or restrict it behind a firewall and proper network configuration—and ensure Erlang Distribution is never exposed to untrusted networks.
+
+## TL;DR
+
+**Check if you’re exposed:**
+- Run a port scan:  
+  ```bash
+  nmap -A -sT -p 4369 [Your Public IP]
+  ```  
+  If the output shows:  
+  ```
+  PORT     STATE SERVICE VERSION
+  4369/tcp open  epmd    Erlang Port Mapper Daemon
+  | epmd-info: 
+  |   epmd_port: 4369
+  |   nodes: 
+  |_    rabbit: 25672
+  ```
+  Your EPMD port is exposed, indicating your Erlang Distribution ports may be discoverable as well.
+
+**For Erlang/Elixir users:**
+
+- **Use a firewall or security group rules to restrict external access**:  
+  Close all ports except those explicitly required for your application. Restricting access to only essential services reduces your overall attack surface and helps mitigate various security risks beyond just EPMD exposure.
+
+- **Bind distribution to a non-public interface**:  
+  In `vm.args`:
+  ```bash
+  -kernel inet_dist_use_interface '{127, 0, 0, 1}'
+  ```
+  Ensuring that distribution ports are only accessible via localhost or a private network interface further reduces the attack surface.
+
+- **Remove or avoid using `-name` or `-sname` if you don’t need clustering**:  
+  If you’re not clustering nodes, don’t assign them a name that registers with EPMD. Without a name, your node won’t appear in EPMD’s registry, making it harder for attackers to locate your distribution ports.
```

Comment:
Removing the `-name`/`-sname` disables distribution completely. It is not only removed from EPMD.